common.go

```diff
@@ -78,8 +93,126 @@ type PaymentMethodDetails struct {
 	AdditionalDetails AdditionalDetails `json:"additional_details,omitempty"`
 }
 
-// PaymentMethodHref wraps PaymentMethod with associated href.
-type PaymentMethodHref struct {
-	Href          string         `json:"href"`
-	PaymentMethod *PaymentMethod `json:"payment_method"`
+// This object represents the configuration of the provider that handled the transaction,
+// as defined in your PaymentsOS Control Center account.
+// Note that the object does not include your provider authentication credentials.
+type ProviderConfiguration struct {
+	ID          string      `json:"id"`
+	Name        string      `json:"name"`
+	Description string      `json:"description"`
+	Created     json.Number `json:"created"`
+	Modified    json.Number `json:"modified"`
+	ProviderID  string      `json:"provider_id"`
+	Type        string      `json:"type"`
+	AccountID   string      `json:"account_id"`
+	Href        string      `json:"href"`
+}
+
+// The line items of the order.
+type OrderLineItem struct {
+	Name      string `json:"name"`
+	ID        string `json:"id"`
+	Quantity  int    `json:"quantity"`
+	UnitPrice int64  `json:"unit_price"`
+}
+
+// Details of the order. Note that order fields required for level 2 and 3 data, must be passed separately
+// in a Create Capture request within a level_2_3 object (fields passed in the order object only are not recognized as level 2 and 3 data fields).
+type Order struct {
+	ID                string            `json:"id"`
+	AdditionalDetails AdditionalDetails `json:"additional_details,omitempty"`
+	LineItems         []OrderLineItem   `json:"line_items"`
+	TaxAmount         int64             `json:"tax_amount"`
+	TaxPercentage     decimal.Decimal   `json:"tax_percentage"`
+}
+
+type DecodedJSON map[string]interface{}
+
+func (d *DecodedJSON) UnmarshalJSON(data []byte) error {
+	var (
+		err           error
+		unquotedData  string
+		rawSet        = make(map[string]json.RawMessage)
+		citationLevel = 0
+	)
+
+	unquotedData, err = strconv.Unquote(string(data))
+	// Error can be returned if it's not quoted string
+	if err != nil {
+		// There are not quoted data, trying to unmarshal
+		err = json.Unmarshal(data, &rawSet)
+	} else {
+		for err == nil && citationLevel < maxCitationLevel {
+			unquotedData, err = strconv.Unquote(string(data))
```

Comment:
Looks like what you meant here is
`unquoutedData, err = strconv.Unquote(unquotedData)`

But then the `unquoutedData` will become empty once `strconv.Unquote()` returns error , so you will have to rewrite the loop.

And the "more more complicated json" test expected outcome is probably wrong.

Reply:
You are right, have fixed it.